Libraries/Opc.Ua.Security.Certificates/Extensions/X509SubjectAltNameExtension.cs

```diff
@@ -0,0 +1,446 @@
+                    sanBuilder.AddIpAddress(ipAddr);
+                }
+                else
+                {
+                    sanBuilder.AddDnsName(generalName);
+                }
+            }
+        }
+#else  
+        /// <summary>
+        /// Encode the Subject Alternative name extension.
+        /// </summary>
+        private byte[] Encode()
+        {
+            return BouncyCastle.X509Extensions.BuildSubjectAltNameExtension(m_uris, m_domainNames, m_ipAddresses).RawData;
+        }
+#endif
+
+        /// <summary>
+        /// Decode if RawData is yet undecoded.
+        /// </summary>
+        private void EnsureDecoded()
+        {
+            if (!m_decoded)
+            {
+                Decode(RawData);
+            }
+        }
+
+        /// <summary>
+        /// Decode URI, DNS and IP from Subject Alternative Name.
+        /// </summary>
+        /// <remarks>
+        /// Only general names relevant for Opc.Ua are decoded.
+        /// </remarks>
+        private void Decode(byte[] data)
+        {
+            if (base.Oid.Value == SubjectAltNameOid ||
+                base.Oid.Value == SubjectAltName2Oid)
+            {
+                try
+                {
+                    List<string> uris = new List<string>();
+                    List<string> domainNames = new List<string>();
+                    List<string> ipAddresses = new List<string>();
+                    Asn1Tag uriTag = new Asn1Tag(TagClass.ContextSpecific, 6);
+                    Asn1Tag dnsTag = new Asn1Tag(TagClass.ContextSpecific, 2);
+                    Asn1Tag ipTag = new Asn1Tag(TagClass.ContextSpecific, 7);
+                    AsnReader dataReader = new AsnReader(data, AsnEncodingRules.DER);
+                    var akiReader = dataReader?.ReadSequence();
+                    if (akiReader != null)
+                    {
+                        while (akiReader.HasData)
+                        {
+                            Asn1Tag peekTag = akiReader.PeekTag();
+                            if (peekTag == uriTag)
+                            {
+                                var uri = akiReader.ReadCharacterString(UniversalTagNumber.IA5String,
+                                    new Asn1Tag(TagClass.ContextSpecific, 6));
+                                uris.Add(uri);
+                            }
+                            else if (peekTag == dnsTag)
+                            {
+                                var dnsName = akiReader.ReadCharacterString(UniversalTagNumber.IA5String,
+                                    new Asn1Tag(TagClass.ContextSpecific, 2));
+                                domainNames.Add(dnsName);
+                            }
+                            else if (peekTag == ipTag)
+                            {
+                                var ip = akiReader.ReadOctetString(new Asn1Tag(TagClass.ContextSpecific, 7));
+                                ipAddresses.Add(IPAddressToString(ip));
```

Comment:
You can reuse the `uriTag`, `dnsTag`, and `ipTag` variables here.  (If this is one of the "but none of the code in the BCL does"-isms, it's because ours is mostly written by a generator, and it's hard to teach generators about variable reuse, but easy to make them write out a new expression to a struct).

Reply:
ok, got it